master/static/migrations/20230509102913_add-generic-metrics-partitioned-table.tx.up.sql

```diff
@@ -0,0 +1,132 @@
+ | trial_run_id  | integer                  |  not null default 0                                     |
+ | archived      | boolean                  |  not null default false                                 |
+ | id            | integer                  |  not null default nextval('raw_steps_id_seq'::regclass) |
+ +---------------+--------------------------+---------------------------------------------------------+
+ Indexes:
+ "steps_trial_id_total_batches_run_id_unique" UNIQUE, btree (trial_id, total_batches, trial_run_id)
+ "steps_archived" btree (archived)
+ Foreign-key constraints:
+ "steps_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(id)
+ 
+ determined> \d raw_validations;
+ +---------------+--------------------------+--------------------------------------------+
+ | Column        | Type                     | Modifiers                                  |
+ |---------------+--------------------------+--------------------------------------------|
+ | id            | integer                  |  not null generated by default as identity |
+ | trial_id      | integer                  |  not null                                  |
+ | end_time      | timestamp with time zone |                                            |
+ | metrics       | jsonb                    |                                            |
+ | total_batches | integer                  |  not null default 0                        |
+ | trial_run_id  | integer                  |  not null default 0                        |
+ | archived      | boolean                  |  not null default false                    |
+ +---------------+--------------------------+--------------------------------------------+
+ Indexes:
+ "validations_pkey" PRIMARY KEY, btree (id)
+ "validations_trial_id_total_batches_run_id_unique" UNIQUE, btree (trial_id, total_batches, trial_run_id)
+ "ix_validations_trial_id" btree (trial_id)
+ "validations_archived" btree (archived)
+ Referenced by:
+ TABLE "trials" CONSTRAINT "trials_latest_validation_id_fkey" FOREIGN KEY (latest_validation_id) REFERENCES raw_validations(id) ON DELETE SET NULL
+ Triggers:
+ autoupdate_exp_validation_metrics_name AFTER INSERT ON raw_validations FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_validation_metrics_name()
+ */
+-- fix missing foreign key constraint
+ALTER TABLE raw_validations
+ADD CONSTRAINT validations_trial_id_fkey FOREIGN KEY (trial_id) REFERENCES trials(id);
+CREATE TYPE metric_partition_type AS ENUM ('validation', 'training', 'generic');
+ALTER TABLE raw_validations
+ADD COLUMN partition_type metric_partition_type NOT NULL DEFAULT 'validation';
+ALTER INDEX validations_trial_id_total_batches_run_id_unique
+RENAME TO validations_trial_id_total_batches_run_id_unique_old;
+CREATE UNIQUE INDEX validations_trial_id_total_batches_run_id_type_unique ON raw_validations (
+    trial_id,
+    total_batches,
+    trial_run_id,
+    partition_type
+);
```

Comment:
could you talk a bit why did you do this? why is `partition_type` necessary in this index?

how does the performance look in general, any impact?

Reply:
write perf looks the same. migration and read testing in progress

Reply:
streaming 180k rows didn't show much difference (per trial limit set to 100k)

```
determined git:(main) ✗ hyperfine --warmup 1 "det dev curl '/api/v1/trials/metrics/training_metrics?trial_ids=10242&trial_ids=10243&trial_ids=10244'"
Benchmark 1: det dev curl '/api/v1/trials/metrics/training_metrics?trial_ids=10242&trial_ids=10243&trial_ids=10244'
  Time (mean ± σ):      5.513 s ±  0.027 s    [User: 0.668 s, System: 0.156 s]
  Range (min … max):    5.484 s …  5.567 s    10 runs
  
```

Reply:
update done.